src/EventStore.Core/TransactionLog/Chunks/TFChunk/WriterWorkItem.cs

```diff
@@ -1,90 +1,74 @@
 
 		public readonly MemoryStream Buffer;
 		public readonly BinaryWriter BufferWriter;
 		public readonly HashAlgorithm MD5;
 
-		public WriterWorkItem(Stream fileStream, UnmanagedMemoryStream memStream, HashAlgorithm md5) {
-			_fileStream = fileStream;
-			_memStream = memStream;
-			_workingStream = (Stream)fileStream ?? memStream;
-			Buffer = new MemoryStream(8192);
-			BufferWriter = new BinaryWriter(Buffer);
+		public unsafe WriterWorkItem(nint memoryPtr, int length, HashAlgorithm md5) {
+			WorkingStream = _memStream = new UnmanagedMemoryStream((byte*)memoryPtr, length, length, FileAccess.ReadWrite);
+			Buffer = new(BufferSize);
+			BufferWriter = new(Buffer);
+			MD5 = md5;
+		}
+
+		public WriterWorkItem(SafeFileHandle handle, HashAlgorithm md5, bool unbuffered) {
+			WorkingStream = _fileStream = unbuffered
+				? handle.AsUnbufferedStream(FileAccess.ReadWrite)
+				: new BufferedStream(handle.AsUnbufferedStream(FileAccess.ReadWrite), BufferSize);
+			Buffer = new(BufferSize);
+			BufferWriter = new(Buffer);
 			MD5 = md5;
 		}
 
 		public void SetMemStream(UnmanagedMemoryStream memStream) {
 			_memStream = memStream;
-			if (_fileStream == null)
-				_workingStream = memStream;
+			if (_fileStream is null)
+				WorkingStream = memStream;
 		}
 
 		public void AppendData(byte[] buf, int offset, int len) {
 			// as we are always append-only, stream's position should be right here
-			if (_fileStream != null)
-				_fileStream.Write(buf, 0, len);
+			_fileStream?.Write(buf, offset, len);
+
 			//MEMORY
-			var memStream = _memStream;
-			if (memStream != null)
-				memStream.Write(buf, 0, len);
+			_memStream?.Write(buf, offset, len);
 		}
 
 		public void ResizeStream(int fileSize) {
-			if (_fileStream != null)
-				_fileStream.SetLength(fileSize);
-			var memStream = _memStream;
-			if (memStream != null)
-				memStream.SetLength(fileSize);
+			_fileStream?.SetLength(fileSize);
+			_memStream?.SetLength(fileSize);
 		}
 
-		public void Dispose() {
-			if (_fileStream != null)
-				_fileStream.Dispose();
+		protected override void Dispose(bool disposing) {
+			if (disposing) {
+				_fileStream?.Dispose();
+				DisposeMemStream();
+				Buffer.Dispose();
+				BufferWriter.Dispose();
+				MD5.Dispose();
+			}
 
-			DisposeMemStream();
+			base.Dispose(disposing);
 		}
 
-		public void FlushToDisk() {
-			if (_fileStream == null) return;
-			var fs = _fileStream as FileStream;
-			if (fs != null)
-				fs.FlushToDisk(); //because of broken flush in filestream in 3.5
-			else {
-				_fileStream.Flush();
-			}
-		}
+		public void FlushToDisk() => _fileStream?.Flush();
 
 		public void DisposeMemStream() {
-			var memStream = _memStream;
-			if (memStream != null) {
-				memStream.Dispose();
-				_memStream = null;
-			}
-		}
-
-		public void EnsureMemStreamLength(long length) {
-			throw new NotSupportedException("Scavenging with in-mem DB is not supported.");
+			_memStream?.Dispose();
+			_memStream = null;
```

Comment:
should `_memStream` be set to null _before_ disposing the memory stream? otherwise it could be possible that we use a disposed stream. (i'm not sure if we need to fix it in this PR as this issue was already present before and it might not be straightforward to fix as it involves adding memory barriers)

Reply:
hmmm, it also doesn't seem to be a problem as we always cache the active chunk now. I wonder if it would be a good idea to get rid of the `DisposeMemStream()` method on `WriterWorkItem`

Reply:
yeah i dont think we need to change anything here, as you say it matches the previous behaviour and it doesn't seem to be a problem anyway